File: init/fs.c
#include "fs.h"

static int try_mount(const char *source, const char *target, const char *fstype,
              unsigned long mountflags, const void *data);

void setup_root_block_device(void)
{
    int fd;
    char *krun_root;
    char *krun_root_fstype;
    char *krun_root_options;

    krun_root = getenv("KRUN_BLOCK_ROOT_DEVICE");
    if (krun_root) {
        if (mkdir("/newroot", 0755) < 0 && errno != EEXIST) {
            perror("mkdir(/newroot)");
            exit(-1);
        }

        krun_root_fstype = getenv("KRUN_BLOCK_ROOT_FSTYPE");
        krun_root_options = getenv("KRUN_BLOCK_ROOT_OPTIONS");

        if (try_mount(krun_root, "/newroot", krun_root_fstype, 0,
                      krun_root_options) < 0) {
            perror("mount KRUN_BLOCK_ROOT_DEVICE");
            exit(-1);
        }

        chdir("/newroot");

        fd = open("/", O_RDONLY);
        if (fd < 0) {
            perror("Couldn't open temporary root directory for removing");
            exit(-1);
        }
        if (ioctl(fd, KRUN_REMOVE_ROOT_DIR_IOCTL) < 0) {
            perror("Error removing temporary root directory");
        }
        close(fd);

        if (mount(".", "/", NULL, MS_MOVE, NULL) < 0) {
            perror("remount root");
            exit(-1);
        }
        chroot(".");

        // we must mount filesystems again after chrooting
        if (mount_filesystems() < 0) {
            printf("Couldn't mount filesystems, bailing out\n");
            exit(-2);
        }
    }

    if (mount(NULL, "/", NULL, MS_REC | MS_SHARED, NULL) < 0) {
        perror("Couldn't set shared propagation on the root mount");
        exit(-1);
    }
}

int mount_filesystems()
{
    char *const DIRS_LEVEL1[] = {"/dev", "/proc", "/sys"};
    char *const DIRS_LEVEL2[] = {"/dev/pts", "/dev/shm"};
    int i;

    for (i = 0; i < 3; ++i) {
        if (mkdir(DIRS_LEVEL1[i], 0755) < 0 && errno != EEXIST) {
            printf("Error creating directory (%s)\n", DIRS_LEVEL1[i]);
            return -1;
        }
    }

    if (mount("devtmpfs", "/dev", "devtmpfs", MS_RELATIME, NULL) < 0 &&
        errno != EBUSY) {
        perror("mount(/dev)");
        return -1;
    }

    if (mount("proc", "/proc", "proc",
              MS_NODEV | MS_NOEXEC | MS_NOSUID | MS_RELATIME, NULL) < 0) {
        perror("mount(/proc)");
        return -1;
    }

    if (mount("sysfs", "/sys", "sysfs",
              MS_NODEV | MS_NOEXEC | MS_NOSUID | MS_RELATIME, NULL) < 0) {
        perror("mount(/sys)");
        return -1;
    }

    if (mount("cgroup2", "/sys/fs/cgroup", "cgroup2",
              MS_NODEV | MS_NOEXEC | MS_NOSUID | MS_RELATIME, NULL) < 0) {
        perror("mount(/sys/fs/cgroup)");
        return -1;
    }

    for (i = 0; i < 2; ++i) {
        if (mkdir(DIRS_LEVEL2[i], 0755) < 0 && errno != EEXIST) {
            printf("Error creating directory (%s)\n", DIRS_LEVEL2[i]);
            return -1;
        }
    }

    if (mount("devpts", "/dev/pts", "devpts",
              MS_NOEXEC | MS_NOSUID | MS_RELATIME, NULL) < 0) {
        perror("mount(/dev/pts)");
        return -1;
    }

    if (mount("tmpfs", "/dev/shm", "tmpfs", MS_NOEXEC | MS_NOSUID | MS_RELATIME,
              NULL) < 0) {
        perror("mount(/dev/shm)");
        return -1;
    }

    /* May fail if already exists and that's fine. */
    symlink("/proc/self/fd", "/dev/fd");

    return 0;
}

static int try_mount(const char *source, const char *target, const char *fstype,
              unsigned long mountflags, const void *data)
{
    FILE *f;
    char line[129];
    int mount_status = -1;

    if (fstype) {
        return mount(source, target, fstype, mountflags, data);
    }

    f = fopen("/proc/filesystems", "r");
    if (f == NULL) {
        perror("fopen(/proc/filesystems)");
        return -1;
    }
    while (fgets(line, sizeof(line), f)) {
        char fstype[sizeof(line)];
        if (!strncmp(line, "nodev", 5)) {
            continue;
        }
        if (sscanf(line, "%128s\n", fstype) != 1) {
            continue;
        }

        mount_status = mount(source, target, fstype, mountflags, data);
        if (mount_status == 0) {
            break;
        }
    }
    fclose(f);

    return mount_status;
}

int is_virtiofs(const char *path)
{
    struct statfs fs;

    if (statfs(path, &fs) != 0) {
        perror("statfs");
        return -1;
    }

    // virtiofs magic number: 0x65735546
    return (fs.f_type == 0x65735546) ? 1 : 0;
}

File: init/src/fs.rs
use std::fs::{self, File};
use std::io::{BufRead, BufReader};
use std::os::unix::fs as unix_fs;

use anyhow::Context;
use nix::errno::Errno;
use nix::mount::{self, MsFlags};
use nix::unistd;

/// Mount, treating EBUSY (already mounted) as success.
fn mount_or_busy(
    src: Option<&str>,
    target: &str,
    fstype: Option<&str>,
    flags: MsFlags,
) -> anyhow::Result<()> {
    match mount::mount(src, target, fstype, flags, None::<&str>) {
        Ok(()) => Ok(()),
        Err(Errno::EBUSY) => Ok(()),
        Err(e) => Err(e).with_context(|| format!("mount {target}")),
    }
}

pub fn mount_filesystems() -> anyhow::Result<()> {
    let base_flags = MsFlags::MS_NOEXEC | MsFlags::MS_NOSUID | MsFlags::MS_RELATIME;
    fs::create_dir_all("/dev").context("create /dev")?;
    fs::create_dir_all("/proc").context("create /proc")?;
    fs::create_dir_all("/sys").context("create /sys")?;

    mount_or_busy(
        Some("devtmpfs"),
        "/dev",
        Some("devtmpfs"),
        MsFlags::MS_RELATIME,
    )?;

    mount_or_busy(
        Some("proc"),
        "/proc",
        Some("proc"),
        MsFlags::MS_NODEV | base_flags,
    )?;

    mount_or_busy(
        Some("sysfs"),
        "/sys",
        Some("sysfs"),
        MsFlags::MS_NODEV | base_flags,
    )?;

    mount_or_busy(
        Some("cgroup2"),
        "/sys/fs/cgroup",
        Some("cgroup2"),
        MsFlags::MS_NODEV | base_flags,
    )?;

    fs::create_dir_all("/dev/pts").context("create /dev/pts")?;
    fs::create_dir_all("/dev/shm").context("create /dev/shm")?;

    mount_or_busy(Some("devpts"), "/dev/pts", Some("devpts"), base_flags)?;
    mount_or_busy(Some("tmpfs"), "/dev/shm", Some("tmpfs"), base_flags)?;

    // Best-effort; may already exist.
    let _ = unix_fs::symlink("/proc/self/fd", "/dev/fd");

    Ok(())
}

/// Returns true if path is listed as a mount point in /proc/mounts.
///
/// Uses /proc/mounts instead of stat() because Podman arranges tmpfs
/// auto-mounts that would be triggered by a stat call.
pub fn is_mount_point(path: &str) -> bool {
    let Ok(f) = File::open("/proc/mounts") else {
        return false;
    };
    for line in BufReader::new(f).lines().map_while(Result::ok) {
        let mut parts = line.split_whitespace();
        let _ = parts.next(); // device
        if parts.next() == Some(path) {
            return true;
        }
    }
    false
}

pub fn mount_tmpfs(path: &str) -> anyhow::Result<()> {
    mount::mount(
        Some("tmpfs"),
        path,
        Some("tmpfs"),
        MsFlags::MS_NOEXEC | MsFlags::MS_NOSUID | MsFlags::MS_NODEV | MsFlags::MS_RELATIME,
        None::<&str>,
    )
    .with_context(|| format!("mount tmpfs at {path}"))
}

/// Mount /dev/vda as ext4, then pivot root into it.
#[cfg(any(feature = "amd-sev", feature = "tdx"))]
pub fn mount_tee_block_device() -> anyhow::Result<()> {
    fs::create_dir_all("/tmp/vda").context("create /tmp/vda")?;

    mount_or_busy(
        Some("/dev/vda"),
        "/tmp/vda",
        Some("ext4"),
        MsFlags::MS_RELATIME,
    )?;
    unistd::chdir("/tmp/vda").context("chdir /tmp/vda")?;

    mount_or_busy(Some("."), "/", None::<&str>, MsFlags::MS_MOVE)?;
    unistd::chroot(".").context("chroot .")
}
